mm2src/coins/qrc20/swap_ops.rs

```diff
@@ -0,0 +1,886 @@
+            UtxoRpcClientEnum::Native(_) => return ERR!("Electrum client expected"),
+        };
+        let qtum_tx: UtxoTx = try_s!(deserialize(verbose_tx.hex.as_slice()).map_err(|e| ERRL!("{:?}", e)));
+
+        // The transaction could not being mined, just check the transfer tokens.
+        let output = qtum_tx
+            .outputs
+            .first()
+            .ok_or(ERRL!("Provided dex fee tx {:?} has no outputs", qtum_tx))?;
+        let script_pubkey: Script = output.script_pubkey.clone().into();
+
+        let (receiver, value) = match transfer_call_details_from_script_pubkey(&script_pubkey) {
+            Ok((rec, val)) => (rec, val),
+            Err(e) => return ERR!("Provided dex fee tx {:?} is incorrect: {}", qtum_tx, e),
+        };
+
+        if receiver != fee_addr {
+            return ERR!(
+                "QRC20 Fee tx was sent to wrong address {:?}, expected {:?}",
+                receiver,
+                fee_addr
+            );
+        }
+
+        if value < expected_value {
+            return ERR!("QRC20 Fee tx value {} is less than expected {}", value, expected_value);
+        }
+
+        let token_addr = try_s!(extract_token_addr_from_script(&script_pubkey));
+        if token_addr != self.contract_address {
+            return ERR!(
+                "QRC20 Fee tx {:?} called wrong smart contract, expected {:?}",
+                qtum_tx,
+                self.contract_address
+            );
+        }
+
+        Ok(())
+    }
+
+    pub async fn search_for_swap_tx_spend(
+        &self,
+        time_lock: u32,
+        secret_hash: Vec<u8>,
+        tx: UtxoTx,
+        search_from_block: u64,
+    ) -> Result<Option<FoundSwapTxSpend>, String> {
+        let electrum = match self.utxo.rpc_client {
+            UtxoRpcClientEnum::Electrum(ref rpc_cln) => rpc_cln,
+
+            UtxoRpcClientEnum::Native(_) => {
+                return ERR!("Native mode not supported");
+            },
+        };
+
+        let tx_hash = tx.hash().reversed().into();
+        let verbose_tx = try_s!(electrum.get_verbose_transaction(tx_hash).compat().await);
+        if verbose_tx.confirmations < 1 {
+            return ERR!("'erc20Payment' was not confirmed yet. Please wait for at least one confirmation");
+        }
+
+        let Erc20PaymentDetails { swap_id, receiver, .. } = try_s!(self.erc20_payment_details_from_tx(&tx).await);
+        let expected_swap_id = qrc20_swap_id(time_lock, &secret_hash);
+        if expected_swap_id != swap_id {
+            return ERR!("Unexpected swap_id {}", hex::encode(swap_id));
+        }
+
+        // First try to find a 'receiverSpend' contract call.
+        // This means that we should request a transaction history for the possible spender of our payment - [`Erc20PaymentDetails::receiver`].
+        let history = try_s!(
+            HistoryBuilder::new(self.clone())
+                .from_block(search_from_block)
```

Comment:
Unfortunately, yes.
Perhaps in native mode it will be possible, I didn't dive into the Native API.
I tried to pass all of the variations of (my_address, token_address, swap_contract_address) into the [blockchain.contract.event.get_history](https://github.com/qtumproject/qtum-electrumx-server/blob/master/docs/qrc20-integration.md#blockchaincontracteventget_historyhash160-contract_addr-topic), also tried to pass different topics (Transfer, PaymentSent).

Electrum returns non-empty history the following combinations:
`0xba8b71f3544b93e2f681f996da519a98ace0107a 0xd362e096e873eb7907e205fadc6175c6fec7bc44 ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef`

`qXxsj5RtciAby9T7m98AgAATL4zTi4UwDG 0xd362e096e873eb7907e205fadc6175c6fec7bc44 ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef`

where 0xba8b71f3544b93e2f681f996da519a98ace0107a - is swap contract address,
0xd362e096e873eb7907e205fadc6175c6fec7bc44 - is token address,
ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef - Transfer topic

Reply:
I will check if we can contact QTUM community in regards to this, maybe they will add the RPC to their ElectrumX fork that allows to fetch arbitrary events from smart contracts, not only QRC20. Native mode should work fine with it I guess. Marking as resolved as we can't do much about it right now.